src/composables/useEventStore/index.ts

```diff
@@ -0,0 +1,104 @@
+import { computed, shallowRef } from 'vue'
+import type { Event, Object3D } from 'three'
+import type { TresContext } from '../useTresContextProvider'
+import { useRaycaster } from '../useRaycaster'
+import { hyphenate } from '../../utils'
+
+export function useEventStore(scene: THREE.Scene, context: TresContext, emit) {
+  const _scene = shallowRef<THREE.Scene>()
+  const _context = shallowRef<TresContext>()
+
+  if (scene) _scene.value = scene
+  if (context) _context.value = context
+
+  const sceneChildren = computed(() => _scene.value ? _scene.value.children : [])
+
+  function executeEventListeners(listeners: Function | Function[], intersection, event) {
+    // Components with multiple event listeners will have an array of functions
+    if (Array.isArray(listeners)) {
+      for (const listener of listeners) {
+        listener(intersection, event)
+      }
+    }
+
+    // Single listener will be a function
+    if (typeof listeners === 'function') {
+      listeners(intersection, event)
+    }
+  }
+
+  /**
+     * propogateEvent
+     * 
+     * Propogates an event to all intersected objects and their parents
+     * @param eventName - The name of the event to propogate
+     * @param event - The event object
+     * @param intersects - An array of intersections
+     */
+  function propogateEvent(eventName: string, event, intersects) {
+    let stopPropagating = false
+    const stopPropagatingFn = () => stopPropagating = true
+    // Loop through all intersected objects and call their event handler
+    if (intersects.length) {
```

Comment:
(Non-blocking)

If you know `intersects` is an array, this conditional can be removed. `for (const v of vs)` won't loop if `vs.length === 0`. So this conditional is already baked into the next line.

Reply:
good catch! Ty, it is removed